pkg/registry/populator.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	// set the bundleimage on the bundle
+	bundle.BundleImage = i.ref
+
+	if err := bundle.AllProvidedAPIsInBundle(); err != nil {
+		return fmt.Errorf("error checking provided apis in bundle %s: %s", bundle.Name, err)
+	}
+
+	bcsv, err := bundle.ClusterServiceVersion()
+	if err != nil {
+		return fmt.Errorf("error getting csv from bundle %s: %s", bundle.Name, err)
+	}
+
+	packageManifest, err := translateAnnotationsIntoPackage(annotationsFile, bcsv)
+	if err != nil {
+		return fmt.Errorf("Could not translate annotations file into packageManifest %s", err)
+	}
+
+	if err := i.loadOperatorBundle(packageManifest, *bundle); err != nil {
+		return fmt.Errorf("Error adding package %s", err)
+	}
+
+	// Finally let's delete all the old bundles
+	if err = i.loader.ClearNonDefaultBundles(packageManifest.PackageName); err != nil {
+		return fmt.Errorf("Error deleting previous bundles: %s", err)
+	}
+
+	return nil
+}
+
+// loadBundle takes the directory that a CSV is in and assumes the rest of the objects in that directory
+// are part of the bundle.
+func loadBundle(csvName string, dir string) (*Bundle, error) {
+	log := logrus.WithFields(logrus.Fields{"dir": dir, "load": "bundle"})
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var errs []error
+	bundle := &Bundle{}
+	for _, f := range files {
+		log = log.WithField("file", f.Name())
+		if f.IsDir() {
+			log.Info("skipping directory")
+			continue
+		}
+
+		if strings.HasPrefix(f.Name(), ".") {
+			log.Info("skipping hidden file")
+			continue
+		}
+
+		log.Info("loading bundle file")
+		path := filepath.Join(dir, f.Name())
+		fileReader, err := os.Open(path)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to load file %s: %s", path, err))
+			continue
+		}
+
+		decoder := yaml.NewYAMLOrJSONDecoder(fileReader, 30)
+		obj := &unstructured.Unstructured{}
+		if err = decoder.Decode(obj); err != nil {
+			logrus.WithError(err).Debugf("could not decode file contents for %s", path)
+			continue
+		}
+
+		// Don't include other CSVs in the bundle
+		if obj.GetKind() == "ClusterServiceVersion" && obj.GetName() != csvName {
+			continue
+		}
+
+		if obj.Object != nil {
+			bundle.Add(obj)
+		}
+	}
+
+	return bundle, utilerrors.NewAggregate(errs)
+}
+
+// findCSV looks through the bundle directory to find a csv
+func (i *DirectoryPopulator) findCSV(manifests string) (*unstructured.Unstructured, error) {
+	log := logrus.WithFields(logrus.Fields{"dir": i.from, "find": "csv"})
+
+	files, err := ioutil.ReadDir(manifests)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read directory %s: %s", manifests, err)
+	}
+
+	var errs []error
+	for _, f := range files {
+		log = log.WithField("file", f.Name())
+		if f.IsDir() {
+			log.Info("skipping directory")
+			continue
+		}
+
+		if strings.HasPrefix(f.Name(), ".") {
+			log.Info("skipping hidden file")
+			continue
+		}
+
+		path := filepath.Join(manifests, f.Name())
+		fileReader, err := os.Open(path)
```

Comment:
Here too. Maybe it's worth splitting out a tiny `func(string, *unstructured.Unstructured)`?